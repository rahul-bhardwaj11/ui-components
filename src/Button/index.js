import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Loader from '../Loader';
import Icon from '../Icon';
import MtButton, {
  BUTTON_SIZES,
  BUTTON_TYPES,
  MT_SIZE_TO_ANT_BUTTON_SIZE_MAP,
  MT_TYPE_ANT_BUTTON_TYPE_MAP,
  BUTTON_LOADER_COLOR
} from './style';

const noop = () => undefined;

class Button extends Component {
  static propTypes = {
    onClick: PropTypes.func,
    type: PropTypes.oneOf(Object.values(BUTTON_TYPES)),
    children: PropTypes.node,
    disabled: PropTypes.bool,
    size: PropTypes.oneOf(Object.values(BUTTON_SIZES)),
    style: PropTypes.object,
    className: PropTypes.string,
    active: PropTypes.bool,
    loading: PropTypes.bool,
    danger: PropTypes.bool
  };
  static defaultProps = {
    onClick: noop,
    children: 'Submit',
    disabled: false,
    type: BUTTON_TYPES.PRIMARY,
    size: 'large',
    danger: false
  };

  render() {
    const {
      children,
      type,
      style = {},
      active,
      disabled,
      size,
      loading,
      danger
    } = this.props;
    let antdType = MT_TYPE_ANT_BUTTON_TYPE_MAP[type];
    let { className, ...rest } = this.props;
    return (
      <MtButton
        {...rest}
        active={active ? 1 : 0}
        disabled={disabled}
        className={className}
        type={antdType}
        style={style}
        danger={danger ? 1 : 0}
        loading={loading}
        size={
          MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[size] ||
          MT_SIZE_TO_ANT_BUTTON_SIZE_MAP['large']
        }
      >
        {loading && (
          <Loader
            size="sizeXSmall"
            type="Full"
            style={{ loaderStyle: { borderColor: BUTTON_LOADER_COLOR[type] } }}
          />
        )}
        {children}
        {type === 'edit' && <Icon type="edit" className="editIcon" />}
      </MtButton>
    );
  }
}
export default Button;
