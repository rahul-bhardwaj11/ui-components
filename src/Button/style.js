import AntButton from 'antd/lib/button';
import 'antd/lib/button/style/index.css';
import styled from 'styled-components';
import theme from '../styles/theme';
import mixins from '../styles/mixins';

export const BUTTON_SIZES = {
  LARGE: 'large',
  SMALL: 'small',
  MEDIUM: 'medium'
};

export const BUTTON_TYPES = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
  LINK: 'link',
  TEXT: 'text',
  EDIT: 'edit'
};

export const BUTTON_LOADER_COLOR = {
  [BUTTON_TYPES.PRIMARY]: theme.colors.WHITE,
  [BUTTON_TYPES.SECONDARY]: theme.colors.DARK_OUTER_SPACE,
  [BUTTON_TYPES.TERTIARY]: theme.colors.OUTER_SPACE,
  [BUTTON_TYPES.LINK]: theme.colors.INDIGO,
  [BUTTON_TYPES.TEXT]: theme.colors.OUTER_SPACE,
  [BUTTON_TYPES.EDIT]: theme.colors.OUTER_SPACE
};

export const MT_SIZE_TO_ANT_BUTTON_SIZE_MAP = {
  [BUTTON_SIZES.LARGE]: BUTTON_SIZES.LARGE,
  [BUTTON_SIZES.MEDIUM]: 'default',
  [BUTTON_SIZES.SMALL]: BUTTON_SIZES.SMALL
};

export const MT_TYPE_ANT_BUTTON_TYPE_MAP = {
  [BUTTON_TYPES.PRIMARY]: 'primary',
  [BUTTON_TYPES.SECONDARY]: 'default',
  [BUTTON_TYPES.TERTIARY]: 'dashed',
  [BUTTON_TYPES.LINK]: 'link',
  [BUTTON_TYPES.TEXT]: 'text',
  [BUTTON_TYPES.EDIT]: 'edit'
};

const BUTTON_TO_SIZE_MAP = {
  [MT_TYPE_ANT_BUTTON_TYPE_MAP[BUTTON_TYPES.PRIMARY]]: {
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.LARGE]]: BUTTON_SIZES.LARGE,
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.MEDIUM]]: 'default',
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.SMALL]]: BUTTON_SIZES.SMALL
  },
  [MT_TYPE_ANT_BUTTON_TYPE_MAP[BUTTON_TYPES.SECONDARY]]: {
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.LARGE]]: BUTTON_SIZES.LARGE,
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.MEDIUM]]: 'default',
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.SMALL]]: BUTTON_SIZES.SMALL
  },
  [MT_TYPE_ANT_BUTTON_TYPE_MAP[BUTTON_TYPES.TERTIARY]]: {
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.LARGE]]: BUTTON_SIZES.LARGE,
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.MEDIUM]]: 'medium',
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.SMALL]]: BUTTON_SIZES.SMALL
  },
  [MT_TYPE_ANT_BUTTON_TYPE_MAP[BUTTON_TYPES.LINK]]: {
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.LARGE]]: BUTTON_SIZES.LARGE,
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.MEDIUM]]: 'medium',
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.SMALL]]: BUTTON_SIZES.SMALL
  },
  [MT_TYPE_ANT_BUTTON_TYPE_MAP[BUTTON_TYPES.TEXT]]: {
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.LARGE]]: BUTTON_SIZES.LARGE,
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.MEDIUM]]: 'medium',
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.SMALL]]: BUTTON_SIZES.SMALL
  },
  [MT_TYPE_ANT_BUTTON_TYPE_MAP[BUTTON_TYPES.EDIT]]: {
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.LARGE]]: BUTTON_SIZES.LARGE,
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.MEDIUM]]: 'medium',
    [MT_SIZE_TO_ANT_BUTTON_SIZE_MAP[BUTTON_SIZES.SMALL]]: BUTTON_SIZES.SMALL
  }
};

export const ANTD_BUTTON_SIZE_PADDING = {
  [BUTTON_SIZES.LARGE]: '0px 24px',
  default: '0px 12px',
  medium: '0px 16px',
  [BUTTON_SIZES.SMALL]: '0px 12px'
};

const MtButton = styled(AntButton)`
  &.ant-btn {
    font-family: inherit;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    border-radius: 4px;
    ${mixins.button()};
    padding: ${props =>
      ANTD_BUTTON_SIZE_PADDING[BUTTON_TO_SIZE_MAP[props.type][props.size]]};

    /*---------- Default Button styles ------------*/
    &.ant-btn-default {
      border: 1px solid ${theme.colors.ALTO};
      color: ${theme.colors.DARK_OUTER_SPACE};

      &.ant-btn-sm {
        font-size: 12px;
      }
      &:hover {
        color: ${theme.colors.DARK_OUTER_SPACE};
        border: 1px solid ${theme.colors.SILVER};
      }
      &:active,
      &:focus,
      &.active {
        border: 1px solid ${theme.colors.INDIGO};
        color: ${theme.colors.INDIGO};
      }
      &.disabled,
      &:disabled,
      &[disabled]:hover {
        border: 1px solid ${theme.colors.DISABLE};
        background-color: ${theme.colors.PORCELAIN};
        color: ${theme.colors.SILVER};
      }
    }

    /*---------- Primary Button styles ------------*/
    &.ant-btn-primary {
      border: 1px solid ${theme.colors.INDIGO};
      background-color: ${theme.colors.INDIGO};
      color: #fff;
      &.ant-btn-sm {
        font-size: 12px;
      }
      &:hover,
      &:focus,
      &:active,
      &.active {
        border: 1px solid ${theme.colors.NEPTUNE};
        background: ${theme.colors.NEPTUNE};
        color: #fff;
      }
      &.disabled,
      &:disabled,
      &[disabled]:hover {
        border: 1px solid ${theme.colors.ALTO};
        background: ${theme.colors.ALTO};
        color: ${theme.colors.WHITE};
      }
    }

    /*---------- Edit Button styles ------------*/
    &.ant-btn-edit {
      border: 1px solid ${theme.colors.ALTO};
      color: ${theme.colors.SILVER};
      &.ant-btn-sm {
        color: ${theme.colors.SILVER};
        font-size: 12px;
      }
      &:hover,
      &:focus,
      &:active,
      &.active {
        border: 1px solid ${theme.colors.ALTO};
        color: ${theme.colors.SILVER};
        span {
          color: ${theme.colors.SILVER};
        }
      }
      &.disabled,
      &:disabled,
      &[disabled]:hover {
        background-color: ${theme.colors.PORCELAIN};
        color: ${theme.colors.SILVER};
        border: 1px solid ${theme.colors.DISABLE};
      }
    }

    /*---------- Dashed Button styles ------------*/
    &.ant-btn-dashed {
      border: 1px dashed ${theme.colors.ALTO};
      background-color: ${theme.colors.WHITE};
      color: ${theme.colors.OUTER_SPACE};

      &.ant-btn-sm {
        font-size: 12px;
      }
      &:hover,
      &:focus,
      &:active,
      &.active {
        color: ${theme.colors.SHARK};
      }
      &.disabled,
      &:disabled,
      &[disabled]:hover {
        border: 1px dashed ${theme.colors.ALTO};
        background-color: ${theme.colors.WHITE};
        color: ${theme.colors.SILVER};
      }
    }
    /*---------- Text Button styles ------------*/
    &.ant-btn-text {
      border: 1px dashed transparent;
      background-color: transparent;
      color: ${props =>
        props.danger
          ? `${theme.colors.BITTERSWEET}`
          : `${theme.colors.OUTER_SPACE}`};

      ${props =>
        props.active
          ? `color: ${theme.colors.INDIGO};
          background-color: ${theme.colors.TROPICAL_BLUE}`
          : null};
      ${props =>
        props.disabled ? `color: ${theme.colors.OUTER_SPACE}` : null};

      &:hover,
      &:focus,
      &:active {
        border: 1px dashed transparent;
        ${props =>
          props.active
            ? `color: ${theme.colors.INDIGO};
            background-color: ${theme.colors.TROPICAL_BLUE}`
            : `background-color: transparent;
              color: ${theme.colors.INDIGO}`};
        ${props =>
          props.danger ? `color: ${theme.colors.BITTERSWEET}` : null};
      }
      &.ant-btn-sm {
        font-size: 12px;
      }
      &.disabled,
      &:disabled,
      &[disabled]:hover {
        border: 1px solid transparent;
        background-color: ${theme.colors.WHITE};
        color: ${theme.colors.SILVER};
      }
    }
    /*---------- Link Blue Button styles ------------*/
    &.ant-btn-link {
      border: 1px solid ${theme.colors.INDIGO};
      background-color: #fff;
      color: ${theme.colors.INDIGO};

      &.ant-btn-sm {
        font-size: 12px;
      }
      &.disabled,
      &:disabled {
        background-color: ${theme.colors.WHITE};
        color: ${theme.colors.SILVER};
        border: 1px solid ${theme.colors.DISABLE};
      }
    }

    &.disabled,
    &:disabled {
      border: 1px solid ${theme.colors.DISABLE};
      color: ${theme.colors.WHITE};
      background: ${theme.colors.ALTO};
    }
  }

  &.ant-btn-clicked {
    outline: none;
    outline-style: none;
  }
  .editIcon {
    padding: 3px;
  }

  /* loading Button styles */

  &.ant-btn-loading:not(.ant-btn-circle):not(.ant-btn-circle-outline):not(.ant-btn-icon-only) {
    &.ant-btn-sm,
    &.ant-btn.lg,
    &.ant-btn {
      padding: ${props =>
        ANTD_BUTTON_SIZE_PADDING[BUTTON_TO_SIZE_MAP[props.type][props.size]]};

      &:before {
        background: transparent;
      }
    }
    .anticon {
      display: none;
    }
  }

  &.ant-btn {
    & > span {
      visibility: ${props => (props.loading ? 'hidden' : 'visible')};
    }
    & > .anticon {
      & + span {
        display: none;
      }
    }
  }
`;

export default MtButton;
