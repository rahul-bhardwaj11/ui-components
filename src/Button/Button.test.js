import React from 'react';
import { shallow } from 'enzyme';
import Button from './index';
import renderer from 'react-test-renderer';

describe('Button', () => {
  var noop = () => undefined;

  it('matches default props snapshot when type is primary', () => {
    const tree = renderer
      .create(
        <Button
          onClick={noop}
          children="Submit"
          disabled={false}
          type="primary"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches default props snapshot when type is default', () => {
    const tree = renderer
      .create(
        <Button
          onClick={noop}
          children="Submit"
          disabled={false}
          type="default"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should handle click', () => {
    const clickSpy = jest.fn();
    const button = shallow(
      <Button onClick={clickSpy} type="primary" children="submit" />
    );
    button.find('Button').simulate('click');
    expect(clickSpy).toHaveBeenCalled();
  });
});
