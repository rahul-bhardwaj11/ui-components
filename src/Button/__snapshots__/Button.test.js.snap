// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches default props snapshot when type is default 1`] = `
<div
  className="sc-bdVaJa eLfJwG"
>
  <button
    className="ant-btn"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span>
      Submit
    </span>
  </button>
</div>
`;

exports[`Button matches default props snapshot when type is primary 1`] = `
<div
  className="sc-bdVaJa eLfJwG"
>
  <button
    className="ant-btn ant-btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span>
      Submit
    </span>
  </button>
</div>
`;
